### PR TITLE
Fix know issues in IPAM controller implementation and add tests

### PR DESCRIPTION
**Pull request by ykulazhenkov**
- fix ip library to correctly handle out of range IPs
- validate IP address in allocator more strictly
- fix check for gataway in config validator
- add tests for IPAM controller

### CODE REVIEW

File: pkg/ipam-controller/allocator/allocator.go [hunk 1]
```diff
@@ -90,37 +90,41 @@ func (pa *poolAllocator) Allocate(ctx context.Context, node string) (nodeAllocat
 	}
 	allocations := pa.getAllocationsAsSlice()
 	var startIP net.IP
-	if len(allocations) == 0 || ip.Distance(ip.Network(pa.cfg.Subnet).IP, allocations[0].StartIP) > 2 {
+	if len(allocations) == 0 || ip.Distance(pa.cfg.Subnet.IP, allocations[0].StartIP) > 2 {
```
Comment (adrianchiris) [inline pkg/ipam-controller/allocator/allocator.go:93]: why did you drop calls to ip.Network() ?

a user may provide a subnet like: 192.168.100.44/24

File: pkg/ipam-controller/allocator/allocator.go [hunk 2]
```diff
@@ -161,8 +165,8 @@ func (pa *poolAllocator) checkAllocation(allocData nodeAllocationInfo) error {
 	// check that StartIP of the range has valid offset.
 	// all ranges have same size, so we can simply check that (StartIP offset - 1) % pa.cfg.PerNodeBlockSize == 0
 	// -1 required because we skip network addressee (e.g. in 192.168.0.0/24, first allocation will be 192.168.0.1)
-	if math.Mod(float64(ip.Distance(ip.Network(pa.cfg.Subnet).IP, allocData.StartIP))-1,
-		float64(pa.cfg.PerNodeBlockSize)) != 0 {
+	distanceFromGW := ip.Distance(pa.cfg.Subnet.IP, allocData.StartIP)
```
Comment (adrianchiris) [inline pkg/ipam-controller/allocator/allocator.go:168]: distanceFromNetworkStart ? 

GW may not bet the .1 address in the network.

Comment (adrianchiris) [reply pkg/ipam-controller/allocator/allocator.go:168]: also see my comment regarding dropping of ip.Network() call

File: pkg/ipam-controller/allocator/allocator.go (hunk 1 above)
Comment (ykulazhenkov) [reply pkg/ipam-controller/allocator/allocator.go:93]: ParseCIDR returns two objects - net.IP and net.IPNet. We use the second one. It contains valid network ip.

File: pkg/ipam-controller/allocator/allocator.go (hunk 2 above)
Comment (ykulazhenkov) [reply pkg/ipam-controller/allocator/allocator.go:168]: Yes, the variable name is incorrect. I will rename it

Comment (ykulazhenkov) [reply pkg/ipam-controller/allocator/allocator.go]: I renamed the variable

**Review by adrianchiris — APPROVED** [review]

**Pull request merged**